### PR TITLE
Calculate dependency counts in memory

### PR DESCRIPTION
**Pull request by badiib**

### CODE REVIEW

File: storage/spanstore/memory/memory.go [hunk 1]
```diff
@@ -51,8 +51,52 @@ func NewStore() *Store {
...
-	// TODO - go through all traces and build dependencies between start and end time
-	return []model.DependencyLink{}, nil
+	deps := map[string]*model.DependencyLink{}
+	m.Lock()
+	defer m.Unlock()
+	startTs := endTs.Add(-1 * lookback)
+	for _, trace := range m.traces {
+		if m.traceIsBetweenStartAndEnd(startTs, endTs, trace) {
+			for _, s := range trace.Spans {
+				parentSpan := m.findSpan(trace, s.ParentSpanID)
```
Comment (yurishkuro) [inline storage/spanstore/memory/memory.go]: remember that in the storage we can have spans with identical span/parent IDs, especially with zipkin-like tracers like jaeger-client-go. I would suggest running each trace through ID deduper first

File: storage/spanstore/memory/memory.go [hunk 2]
```diff
@@ -46,13 +48,60 @@ func NewStore() *Store {
...
+	defer m.Unlock()
+	startTs := endTs.Add(-1 * lookback)
+	for _, orig := range m.traces {
+		// SpanIDDeduper never returns an err
+		trace, _ := m.deduper.Adjust(orig)
+		if m.traceIsBetweenStartAndEnd(startTs, endTs, trace) {
+			for _, s := range trace.Spans {
+				parentSpan := m.findSpan(trace, s.ParentSpanID)
+				if parentSpan != nil {
+					depKey := parentSpan.Process.ServiceName + "&&&" + s.Process.ServiceName
```
Comment (yurishkuro) [inline storage/spanstore/memory/memory.go]: I would skip the count if `parentSpan.Process.ServiceName == s.Process.ServiceName`?

File: storage/spanstore/memory/memory.go [hunk 3]
```diff
@@ -46,13 +48,60 @@ func NewStore() *Store {
...
+	}
+	retMe := make([]model.DependencyLink, 0, len(deps))
+	for _, dep := range deps {
+		retMe = append(retMe, *dep)
+	}
+	return retMe, nil
+}
+
+func (m *Store) findSpan(trace *model.Trace, spanID model.SpanID) *model.Span {
+	for _, s := range trace.Spans {
```
Comment (yurishkuro) [inline storage/spanstore/memory/memory.go]: trace has FindBySpanID function

File: storage/spanstore/memory/memory.go [hunk 4]
```diff
@@ -46,13 +48,60 @@ func NewStore() *Store {
...
+		if s.SpanID == spanID {
+			return s
+		}
+	}
+	return nil
+}
+
+func (m *Store) traceIsBetweenStartAndEnd(startTs, endTs time.Time, trace *model.Trace) bool {
+	for _, s := range trace.Spans {
+		if s.StartTime.After(startTs) && endTs.After(s.StartTime) {
```
Comment (yurishkuro) [inline storage/spanstore/memory/memory.go]: the interval is exclusive, so in theory you can miss traces. You should use `!After` for one of the conditions to make the interval open/closed.

File: storage/spanstore/memory/memory.go [hunk 5]
```diff
@@ -46,13 +48,60 @@ func NewStore() *Store {
...
-	// TODO - go through all traces and build dependencies between start and end time
-	return []model.DependencyLink{}, nil
+	deps := map[string]*model.DependencyLink{}
+	m.Lock()
+	defer m.Unlock()
+	startTs := endTs.Add(-1 * lookback)
+	for _, orig := range m.traces {
+		// SpanIDDeduper never returns an err
+		trace, _ := m.deduper.Adjust(orig)
+		if m.traceIsBetweenStartAndEnd(startTs, endTs, trace) {
```
Comment (black-adder) [inline storage/spanstore/memory/memory.go]: nit: reduce the level of indentation
```
if !m.traceIsBetweenStartAndEnd(startTs, endTs, trace) {
    continue
}
```

File: storage/spanstore/memory/memory.go [hunk 6]
```diff
@@ -46,13 +48,60 @@ func NewStore() *Store {
...
+			for _, s := range trace.Spans {
+				parentSpan := m.findSpan(trace, s.ParentSpanID)
```
Comment (black-adder) [inline storage/spanstore/memory/memory.go]: nit: reduce level of indentation

```
parentSpan := m.findSpan(trace, s.ParentSpanID)
if parentSpan == nil {
	continue
}
```

**Pull request merged**